onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -298,6 +311,17 @@ Status CUDAExecutionProvider::OnRunEnd() {
   auto current_deferred_release_event = GetPerThreadContext().GetCurrentDeferredReleaseEvent();
   CUDA_RETURN_IF_ERROR(cudaEventRecord(current_deferred_release_event, static_cast<cudaStream_t>(GetComputeStream())));
   CUDA_RETURN_IF_ERROR(cudaStreamSynchronize(static_cast<cudaStream_t>(GetComputeStream())));
+
+  if (shrink_default_memory_allocator_on_every_run_end_) {
+    // We currently do not support shrinking other allocators this EP may hold (like CUDA_PINNED, CUDA_CPU).
+    auto default_memory_alloc = GetAllocator(info_.device_id, OrtMemTypeDefault);
+
+    // NOT 1: We have already ensured that the default memory allocator for the CUDA EP is an arena based allocator
+    // by this point if shrink_default_memory_allocator_on_every_run_end_ is set to true. o the following cast is safe.
+    // NOTE 2: Shrink is thread-safe and the call without any additional synchronization primitives is safe.
+    static_cast<IArenaAllocator*>(default_memory_alloc.get())->Shrink();
+  }
+
```

Comment:
This setup feels a little ugly. You have a setting in an arena about shrinking, that needs to be aware of what a run is, that is triggered via an EP.

Naively I would have expected that this was triggered from InferenceSession::Run, and that it simply iterated the allocators in SessionState calling Shrink for each. Possibly an allocator could have Shrink enabled/disabled if desired.

Reply:
The arena does not need to know what a Run is- it exposes an interface for shrinking - that's all. It is upto the caller to decide whether to want to invoke it / and if they do at what cadence they want to invoke it (the cadence may be at the end of every Run or some other time frequency). 

The EP already has an interface to do some stateful cleanup at the end of every Run (OnRunEnd()) and so the concept of a Run is nothing new to an EP. I make the shrink call from the EP itself instead of making the call from the InferenceSession's Run because we have to check if the default memory arena of each EP of the session is configured for shrinking in the first place - I keep all these checking logic in the EP itself as I feel it is more cleaner to do such checks here than in InferenceSession's Run.

Reply:
Unfortunately we created a bit of a tangled mess by fairly tightly coupling allocators to EPs instead of to devices. 

IIRC we recently allowed sharing the CUDA allocator across the CUDA and TRT EPs, so the allocator isn't necessarily 1:1 with an EP. It would make sense for all the CPU based EPs to share a single allocator instance as well.

Due to that, triggering the Shrink call from the EP seems off as you'll potentially get multiple calls to Shrink for a single allocator if multiple EPs use it. Whilst that doesn't hurt per-se, it's a somewhat messy state to be in. 

If you use the registered allocators in SessionState you're getting one instance per OrtMemoryInfo (we only keep the first registration so the allocator from the highest priority EP wins), and know that you're talking to the same allocators that the ExecutionFrame used.

Reply:
Fixed based on offline sync